[WIP] Sync halo2 lib 0.4.0

zhenfeizhang opened the pull request.

han0110 committed:
feat: call synthesize in `MockProver` multiple times to behave same as real prover

han0110 committed:
Expose mod `permutation` and re-export `permutation::keygen::Assembly` (#149)

* feat: expose mod ule `permutation` and re-export `permutation::keygen::Assembly`

* feat: derive `lone` for `permutation::keygen::Assembly`

* feat: bump MSRV for `inferno`

CPerezz committed:
change: Migrate workspace to pasta_curves-0.5 (#157)

* change: Migrate workspace to pasta_curves-0.5

This ports the majority of the workspace to the `pasta_curves-0.5.0`
leaving some tricky edge-cases that we need to handle carefully.

Resolves: #132

* fix: Complete latest trait bounds to compile halo2proofs

* change: Migrate examples & benches to pasta 0.5

* change: Migrate halo2_gadgets to pasta-0.5

* change: Update gadgets outdated code with latest upstream

* fix: Sha3 gadget circuit

* fix: doc tests

* chore: Update merged main

* fix: Apply review suggestions

ed255 committed:
Extend Circuit trait to take parameters in config (#168)

* Extend Circuit trait to take parameters in config

The Circuit trait is extended with the following:
```
pub trait Circuit<F: Field> {
    /// [...]
    type Params: Default;

    fn params(&self) -> Self::Params {
        Self::Params::default()
    }

    fn configure_with_params(meta: &mut ConstraintSystem<F>, params: &Self::Params) -> Self::Config {
        Self::configure(meta)
    }

    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config;
}
```

This allows runtime parametrization of the circuit configuration.  The extension to the Circuit trait has been designed to minimize the breaking change: existing circuits only need to define the associated `type Params`.

Unfortunately "Associated type defaults" are unstable in Rust, otherwise this would be a non-breaking change.  See https://github.com/rust-lang/rust/issues/29661

* Implement circuit params under feature flag

* Don't overwrite configure method

* Fix doc test

CeciliaZ030 committed:
Allow halo2 constraint names to have non static names (#156)

* static ref to String type in Gates, Constraints, VirtualCell, Argument

* 'lookup'.to_string()

* return &str for gate name and constriant_name, also run fmt

* Update halo2_gadgets/Cargo.toml

Co-authored-by: Han <tinghan0110@gmail.com>

* upgrade rust-toochain

---------

Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>
Co-authored-by: Han <tinghan0110@gmail.com>

CeciliaZ030 committed:
Improve halo2 query calls (#154)

* return expression from cell

* add example

* selector

* recurse Expression to fill in index

* minimized changes from the original

* backword compatible meta.query_X & challange.expr()

* cargo fmt

* fixed lookup to pass all tests

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* Update comments

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* update

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>

* add primitives.rs back

* remove example2

* backward compatible meta.query_X & Column.cur(), next(), prev(), at(usize)

* impl Debug & make side effects only when query.index.is_none()

* change impl Debug for Expression instead & revert test in plonk_api

* upgrade rust-toolchain

* Update halo2_proofs/src/plonk/circuit.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* Update halo2_proofs/src/plonk/circuit.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* ran clippy

* Update halo2_proofs/src/plonk/circuit.rs

Co-authored-by: Han <tinghan0110@gmail.com>

---------

Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>
Co-authored-by: Han <tinghan0110@gmail.com>

enricobottazzi committed:
Implement Clone trait for Hash, Absorbing, and Sponge structs (#171)

CPerezz committed:
fix: Fix serialization for VerifyingKey (#178)

Now the value returned when the number of selectors is a multiple of 8
is correct.

Resolves: #175

ed255 committed:
Add more getters to expose internal fields

Ethan-000 committed:
add a constructor (#164)

* add a constructor

* add more comment

* fix as review

* remove clone

* remove

* no need to use new variable

* change comment

* fix clippy

* rename to from_parts

* remove n declaration

kilic committed:
implement native shuffle argument and api

fix: remove nonsense comment

strictly check shuffle rows

address doc typos

move compression into product commitment

typo

add shuffle errors for `verify_at_rows_par`

dedup expression evaluation

cargo fmt

fix fields in sanity-checks feature

alexander-camuto committed:
feat: send sync region (#180)

* feat: send / sync region

* Update layout.rs

* update

* lol

* debug

* Update keygen.rs

* Update keygen.rs

* Update keygen.rs

* Update keygen.rs

* thread-safe-region feature flag

* cleanup

* patch dev-graph

* patch non-determinism in mapping creation

* reduce mem usage for vk and pk

* mock proving examples

* swap for hashmap for insertion speed

* reduce update overhead

* replace BTree with Vec

* add benchmarks

* make the benchmarks massive

* patch clippy

* simplify lifetimes

* patch benches

* Update halo2_proofs/src/plonk/permutation/keygen.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* Update halo2_proofs/examples/vector-mul.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* rm benches

* order once

* patch lints

---------

Co-authored-by: Han <tinghan0110@gmail.com>

mratsim committed:
Fix `parallelize` workload imbalance (#186)

* fix parallelize workload imbalance

* remove the need of unsafe

huitseeker committed:
Updates halo2_curves dependency to released package (#190)

THe package release ressets the version from those inherited by the legacy
halo2curves repo's fork history.

The upstream diff is:
https://github.com/privacy-scaling-explorations/halo2curves/compare/9f5c50810bbefe779ee5cf1d852b2fe85dc35d5e..9a7f726fa74c8765bc7cdab11519cf285d169ecf

alexander-camuto committed:
feat: public cells to allow for implementations of custom `Layouter`  (#192)

* feat: public cells

* Update mds.rs

* Update mds.rs

* Update single_pass.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* bump toolchain to resolve errors

* fix clippy errors for CI run

* rustfmt post clippy

* plz let it be the last lint

* patch clippy lints in gadgets

* clippy lints for sha256 bench

* patch halo2proof benches

* Update assigned.rs

* Update halo2_gadgets/src/poseidon/primitives/mds.rs

Co-authored-by: Han <tinghan0110@gmail.com>

* Update halo2_gadgets/src/poseidon/primitives/mds.rs

Co-authored-by: Han <tinghan0110@gmail.com>

---------

Co-authored-by: Han <tinghan0110@gmail.com>

alexander-camuto committed:
fix: explicitly define mds diff type (#196)

* fix: explicitly define mds diff type

* rm paren

han0110 committed:
feat: expose `transcript_repr` of `VerifyingKey` and reduce the trait constraint (#200)

han0110 committed:
Synchronize with upstream (#199)

* refactor: add default impl for `SyncDeps` for backward compatability

* feat: pick changes from https://github.com/zcash/halo2/pull/728 and changes of flag `test-dev-graph`

* feat: pick changes from https://github.com/zcash/halo2/pull/622

* feat: pick changes about mod `circuit` and mod `dev`

* feat: pick rest changes of `halo2_proofs`

* fix: when `--no-default-features`

* ci: sync from upstream, and deduplicate jobs when
push to `main`, and remove always failing job `codecov`.

* fix: make `commit_zk` runnable when `--no-default-features`

CPerezz committed:
chore: Update rust-toolchain to 1.66 for testing  (#208)

* chore: Update rust-toolchain to 1.66 for testing

Note that tests will not compile due to the silent MSRV bump in
`blake2b_simd`.

Hence, we need to use `1.66` as toolchain.

Resolves: #207

* change: UIpdate MSRVs in Cargo.toml

thor314 committed:
fix: clippy (#203)

* fix: clippy

* fmt

* fix: Final clippy complains & adjustments

---------

Co-authored-by: CPerezz <c.perezbaro@gmail.com>

pnyda committed:
Implement Sum and Product for Expression (#209)

* Make it Eq to make it easier for tests

* Implement Sum and Product for Expression

* Make it readable

Velaciela committed:
modify previous commit

Velaciela committed:
fix previous commit

Velaciela committed:
fix previous commit

zhenfeizhang committed:
chore: update poseidon dependency

zhenfeizhang committed:
fix: compiling bug with feautes=parallel_syn

Velaciela committed:
feat(MockProver): replace errors by asserts(privacy-scaling-explorations#150)

Velaciela committed:
boundary offset lost when resolving conflict

Velaciela commented:
## Change Log

### New feature

- [PSE#129](https://github.com/privacy-scaling-explorations/halo2/pull/129):  Call `synthesize` in `MockProver` multiple times to behave same as real prover
&nbsp;&nbsp;&nbsp;&nbsp; related to scroll/halo2: [#61](https://github.com/scroll-tech/halo2/pull/61), and commit [7a62a6bd](https://github.com/scroll-tech/halo2/commit/7a62a6bdad8c7c878d4c9c4eaa4713a4d898b9ba): add phase-check feature
- [PSE#168](https://github.com/privacy-scaling-explorations/halo2/pull/168): Extend Circuit trait to take parameters in config
&nbsp;&nbsp;&nbsp;&nbsp; feature flag `circuit-params` and type Params = ();
- [PSE#180](https://github.com/privacy-scaling-explorations/halo2/pull/180): feat: send sync region
&nbsp;&nbsp;&nbsp;&nbsp; To enable parallel layouts for large circuits, the Region struct is modified to be Send and Sync by constraining its inner element
- [PSE#185](https://github.com/privacy-scaling-explorations/halo2/pull/185): Implement native shuffle argument and api
&nbsp;&nbsp;&nbsp;&nbsp; introduces ConstraintSystem::shuffle API, enabling efficient one-to-one mapping in dynamic lookup applications, saving permutation witness columns
- [PSE#209](https://github.com/privacy-scaling-explorations/halo2/pull/209): Implement Sum & Prod trait impls for Expression
    
### Improvement      
  
- [PSE#150](https://github.com/privacy-scaling-explorations/halo2/pull/150): feat(MockProver): replace errors by asserts
&nbsp;&nbsp;&nbsp;&nbsp; In MockProver, replace all code that returns an error by an assert that panics instead of returning the error
- [PSE#156](https://github.com/privacy-scaling-explorations/halo2/pull/156): Allow halo2 constraint names to have non static names
&nbsp;&nbsp;&nbsp;&nbsp; `name: &'static str`  >>> `name: AsRef<str>`
- [PSE#154](https://github.com/privacy-scaling-explorations/halo2/pull/154): Improve halo2 query calls
&nbsp;&nbsp;&nbsp;&nbsp; Remove the dependency on VirtualCells to query cells from columns
- [PSE#186](https://github.com/privacy-scaling-explorations/halo2/pull/186): Fix parallelize workload imbalance
&nbsp;&nbsp;&nbsp;&nbsp; performance: arithmetic.rs \ parallelize()
- [PSE#192](https://github.com/privacy-scaling-explorations/halo2/pull/192): feat: public cells to allow for implementations of custom Layouter
&nbsp;&nbsp;&nbsp;&nbsp; exposes the Layouter trait to allow for custom layouter implementation and makes cell.region_index and cell.row_offset public
- [PSE#199](https://github.com/privacy-scaling-explorations/halo2/pull/199): Synchronize with zcash-halo2 [+2024, −1752]
&nbsp;&nbsp;&nbsp;&nbsp; updates the impl for SyncDeps to maintain backward compatibility, modifies CI to test different feature sets, resolves issues
- [PSE#211](https://github.com/privacy-scaling-explorations/halo2/pull/211): Improve root calculation for KZG setup()
&nbsp;&nbsp;&nbsp;&nbsp; performance: ParamsKZG::setup

### API change

- [PSE#129](https://github.com/privacy-scaling-explorations/halo2/pull/149): Expose mod permutation and re-export permutation::keygen::Assembly
- [PSE#157](https://github.com/privacy-scaling-explorations/halo2/pull/157): change: Migrate workspace to pasta_curves-0.5
&nbsp;&nbsp;&nbsp;&nbsp; ports the the workspace to the pasta_curves-0.5.0, ff-0.13 and group-0.13 deps
&nbsp;&nbsp;&nbsp;&nbsp; also to update a bit the halo2-gadgets
- [PSE#181](https://github.com/privacy-scaling-explorations/halo2/pull/181): Add more getters to expose internal fields
- [PSE#164](https://github.com/privacy-scaling-explorations/halo2/pull/164): Adding a constructor for ParamsKZG

### Bug fix

- [PSE#178](https://github.com/privacy-scaling-explorations/halo2/pull/178): fix: Fix serialization for VerifyingKey
&nbsp;&nbsp;&nbsp;&nbsp; `selector.len() / 8 + 1` >>> `(selector.len() + 7) / 8`

### Commits that we didn't pick

- [PSE#152:](https://github.com/privacy-scaling-explorations/halo2/pull/152) Parallel random blinder poly impl 
&nbsp;&nbsp;&nbsp;&nbsp; we use [`domain.constant_lagrange(C::Scalar::one());`](https://github.com/scroll-tech/halo2/blob/92fe9b3e2420ee51336b39513e397df7c8175cf9/halo2_proofs/src/plonk/vanishing/prover.rs#L52-L56) as random_poly

zhenfeizhang committed:
disable multiphase prover

lispc committed:
Sync halo2 lib 0.4.0 merging (#81)

* Use thread pool for assign_regions (#57)

* feat: use rayon threadpool

* feat: add UT for many subregions

* refact: move common struct out to module level

* refact: reuse common configure code

* fix ci errors

---------

Co-authored-by: kunxian xia <xiakunxian130@gmail.com>

* Move `env_logger` dependency to dev-depdendencies (only for test). (#69)

* sync ff/group 0.13

* fix clippy

* fix clippy

* fmg

* [FEAT] Upgrading table16 for SHA256 (#73)

* upgrade sha256

* fix clippy

* Bus auto (#72)

* bus: expose global offset of regions

* bus-auto: add query_advice and query_fixed function in witness generation

* bus-auto: fix clippy

---------

Co-authored-by: Aurélien Nicolas <info@nau.re>

* fix-tob-scroll-21 (#59)

* fix-tob-scroll-21

* expose param field for re-randomization

* enable accessing for table16 (#75)

* chore: update poseidon link

* merge sha256 gadget changes

* Fix the CI errors (#78)

* cargo fmt

* fix clippy error

* Feat: switch to logup scheme for lookup argument  (#71)

* Multi-input mv-lookup. (#49)

* Add mv_lookup.rs

* mv_lookup::prover, mv_lookup::verifier

* Replace lookup with mv_lookup

* replace halo2 with mv lookup

Co-authored-by: ying tong <therealyingtong@users.noreply.github.com>

* cleanups

Co-authored-by: ying tong <therealyingtong@users.noreply.github.com>

* ConstraintSystem: setup lookup_tracker

Co-authored-by: Andrija <akinovak@gmail.com>

* mv_lookup::hybrid_prover

Co-authored-by: Andrija <akinovak@gmail.com>

* WIP

* mv_multi_lookup: enable lookup caching

Co-authored-by: therealyingtong <yingtong.lai@gmail.com>

* Rename hybrid_lookup -> lookup

* Chunk lookups using user-provided minimum degree

Co-authored-by: Andrija <akinovak@gmail.com>

* mv_lookup bench

Co-authored-by: Andrija <akinovak@gmail.com>

* Introduce counter feature for FFTs and MSMs

Co-authored-by: Andrija <akinovak@gmail.com>

* Fix off-by-one errors in chunk_lookup

Co-authored-by: Andrija <akinovak@gmail.com>

* bench wip

* time evaluate_h

* KZG

* more efficient batch inversion

* extended lookup example

* Finalize mv lookup

Author: therealyingtong <yingtong.lai@gmail.com>

* Remove main/

* Fix according to the comments

* replace scan with parallel grand sum computation

* Revert Cargo.lock

* mv lookup Argument name

* parallel batch invert

---------

Co-authored-by: Andrija <akinovak@gmail.com>
Co-authored-by: ying tong <therealyingtong@users.noreply.github.com>
Co-authored-by: therealyingtong <yingtong.lai@gmail.com>

* fmt

* fix unit test

* fix clippy errors

* add todo in mv_lookup's prover

* fmt and clippy

* fix clippy

* add detailed running time of steps in logup's prover

* fmt

* add more log hooks

* more running time logs

* use par invert

* use sorted-vector to store how many times a table element occurs in input

* par the process to get inputs_inv_sum

* use par

* fix par

* add feature to skip inv sums

* add new feature flag

* fix clippy error

---------

Co-authored-by: Sphere L <sph6r6.l1u@gmail.com>
Co-authored-by: Andrija <akinovak@gmail.com>
Co-authored-by: ying tong <therealyingtong@users.noreply.github.com>
Co-authored-by: therealyingtong <yingtong.lai@gmail.com>

* fix some simple building errs

* upgrade pathfinder_simd to newer version as it can't compile on mac m1 pro

* resolve merge conflict

* fmt

* clippy

* more clippy fix

* more lint fix

* fmt

* minor syntax fix

* fix ipa multiopen test failure

* fix clippy warning

* fmt

* fix par scan of log_inv diff

* remove uncessary clone

---------

Co-authored-by: alannotnerd <alan1995wang@outlook.com>
Co-authored-by: kunxian xia <xiakunxian130@gmail.com>
Co-authored-by: Steven <asongala@163.com>
Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>
Co-authored-by: zhenfei <zhenfei.zhang@hotmail.com>
Co-authored-by: Ho <noel.wei@gmail.com>
Co-authored-by: naure <naure@users.noreply.github.com>
Co-authored-by: Aurélien Nicolas <info@nau.re>
Co-authored-by: Sphere L <sph6r6.l1u@gmail.com>
Co-authored-by: Andrija <akinovak@gmail.com>
Co-authored-by: ying tong <therealyingtong@users.noreply.github.com>
Co-authored-by: therealyingtong <yingtong.lai@gmail.com>

kunxian-xia committed:
Merge branch 'v1.1' into sync-halo2-lib-0.4.0
